### ISSUE TITLE
fix(logup_gkr): scan_build_jagged_pyramid stacks all layers at floor width — runtime host-RAM blowup (~100GB) OOMs even a fibonacci prove

### ISSUE THREAD

**Issue by junbeomlee** [fractalyze/zorch#362]
### Summary

`zorch.logup_gkr.circuit.scan_build_jagged_pyramid` builds the LogUp-GKR pyramid on a **fixed-width** `lax.scan`: every one of the ~N pyramid layers is **padded to the floor (widest) layer's width and stacked**, so peak memory is `O(N × plane_width)` instead of the `O(plane_width)` a geometrically-shrinking pyramid should cost. At prove time this stacked intermediate overflows the 32 GB GPU and lands on the **host** — a single *fibonacci* shard reaches **~100 GB host RSS** and OOM-kills the box.

Discovered while bringing up live proving on `fractalyze/sp1-zorch` (the whir-zorch→sp1-zorch cutover, `fractalyze/zkx-run#61`). Filing here because the builder and the fix are in `zorch` (scheme-agnostic).

### Affected version

`zorch` published wheel paired with `sp1_zorch 0.1.0.dev20260702021130`, jax `0.10.0.dev20260701144444`. Repro: fibonacci `elf→core→…→shrink` live e2e (`fractalyze/sp1-zorch` `libs/zkvm/sp1/tests/test_pipeline_e2e.py`).

### Root cause

`prove_logup_gkr_body` calls `scan_build_jagged_pyramid` (`zorch/logup_gkr/circuit.py`):

```python
plane_width = max(sum(counts) for counts in prepad_counts)   # = the FLOOR (widest) layer
init = (_pad_to_width(first.numerator_0, plane_width, 0), ... 4 EF planes ...)
# every plane padded to plane_width, stacked over all ~N transitions as the scan xs/output
```

The fixed-width scan carry + stacked `xs` keep all ~N layers resident at the widest width (4 EF planes × `plane_width` × N × 16 B, koalabearx4). The retired rolled `prove_jagged_pyramid` kept "at most one big layer live"; the fused scan build (wired by `fractalyze/sp1-zorch#143`) lost that invariant.

### Bisection (measured this session, fibonacci shard)

Per-round host-RSS probes (jax.live_arrays split host/device + process RSS + `nvidia-smi`), on the wheel:

| Stage | host RSS | verdict |
|---|---|---|
| riscv-witness **tracegen** (prove skipped) | 1.3 GB | innocent — trace is ~20 MB; chips at natural sizes (max `sp1_range` 2¹⁷, `sp1_global` 13440×241), **not** 2²²-padded |
| **TraceCommitRound** | 1.9 → 6.9 GB | innocent (+5 GB) |
| **LogupGkrRound** | 6.9 → ~100 GB → **OOM** | **culprit** |
| ↳ `resolve_witness_and_grind` (grind) | 6.95 → 6.96 GB, 13 s | innocent (+0.01 GB) |
| ↳ `prove_logup_gkr_body` → `scan_build_jagged_pyramid` | balloons → OOM | **culprit** |

Three properties pin it down:

- **Runtime, not compile-time.** A/B with a warm `JAX_COMPILATION_CACHE_DIR` (146 "Persistent compilation cache hit"; real compile skipped) still peaked **104 GB** and OOM'd, identical to the cold run (107 GB). The balloon is *execution*.
- **Host, not device.** ~100 GB is `RssAnon` (native heap; `RssFile` ≈ 0.5 GB) while the GPU holds ~24 GB at ~0 % util — the geometrically-oversized scan intermediate is landing on host.
- **Grind-independent.** With a witness supplied so the grind never runs (`resolve_witness_and_grind` early-returns — the grind never executes), the body still balloons (the 1 s sampler caught ≥71 GB before the kill) and OOM-kills the box. The memory is pre-existing, not from `fractalyze/sp1-zorch#199`'s new grinder — that PR only *exposed* it (before it, live proving raised `grinding not built` in `resolve_witness_and_grind`, which sits *before* the body, so the pyramid was never reached on the live path).

### What's new vs the existing cluster

`fractalyze/sp1-zorch#55` and `fractalyze/sp1-zorch#186` measured this on **rsp shard17 (1.17 M rows)** and framed it as **compile-time** host RAM / GPU-OOM. This report adds that it is **(a) runtime** (survives a warm compile cache), **(b) structural on a tiny fibonacci trace** (~20 MB → ~100 GB, not shape-scaling), and **(c) on the live prove path** now reachable via `fractalyze/sp1-zorch#199`. `fractalyze/sp1-zorch#143` wired `scan_build_jagged_pyramid` to cut *dispatch count*; it did not address the builder's memory — this is that memory sequel.

### Proposed fix

Make the builder's peak `O(plane_width)` rather than `O(N × plane_width)` while keeping `fractalyze/sp1-zorch#143`'s fused-dispatch / D2H-sync win:

- Don't pad every layer to `plane_width`; segment the scan by width tiers so each tier rides its own (shrinking) buffer, **or**
- build-and-consume layer-by-layer, releasing each proved layer (restore the rolled "one big layer live" invariant), **or**
- a rematerialize/`donate` policy so at most one full-width layer is resident.

### Impact

Blocks `fractalyze/zkx-run#61` (live cutover) and any live/standalone sp1-zorch prove: a fibonacci shard cannot prove on a 123 GB host, and rsp shards are far larger. cf. `fractalyze/sp1-zorch#55`, `fractalyze/sp1-zorch#143`, `fractalyze/sp1-zorch#186`, `fractalyze/sp1-zorch#119`.

**Comment by junbeomlee** [comment]
### Update: markless (host-FS, `jit=False`) does NOT OOM — peak ~37 GB vs ~104 GB

Re-ran the same fibonacci prove with the LogUp-GKR round run **markless / host-orchestrated** (`jit=False` on the shard chain — the host-FS per-round exported-dispatch path, not staged into one fused program):

| mode | peak host RSS | result |
|---|---|---|
| **marked** (`jit=True`, one fused program) | ~104 GB | OOM |
| **markless** (`jit=False`, host-orchestrated per-round) | **~37.5 GB** (plateau ~32 GB) | **no OOM** (timed out at 900 s — dispatch-bound) |

So the ~100 GB balloon is not `scan_build_jagged_pyramid`'s stacked output *alone* (that fits — the markless run still calls `scan_build` and peaks ~32 GB). It is the **`jit=True` whole-body fusion keeping all pyramid layers + all per-round intermediates simultaneously live** (device-side, overflowing the 32 GB card onto host). Running the body host-orchestrated (eager per-round dispatch, FS on host) frees intermediates between rounds → peak drops to ~37 GB and fits.

Caveat: markless traded the OOM for severe slowness — it did **not finish** in 900 s (dispatch-bound / GPU-idle, cf. fractalyze/sp1-zorch#119 and zorch#334). So markless is a memory-viable but slow fallback, not a free fix. The real fix is a **memory-bounded fusion** of the body: keep #143's fused-dispatch win but don't hold all layers live at once (width-tiered scan / build-consume-release per layer / rematerialize), so peak is `O(floor)` on a single device without falling back to the slow host-orchestrated path.

(External `/proc` RSS + `nvidia-smi` sampler; actor stdout was lost to buffering + the timeout kill, so the per-round `ROUNDMEM` prints aren't in this run's log — the sampler curve is the source.)

**Comment by junbeomlee** [comment]
### Correction: markless is NOT memory-bounded — it leaks host RAM monotonically

My earlier comment reported markless (`jit=False`, host-FS per-round dispatch) as "peak ~37 GB, no OOM". **That was an artifact of short timeouts cutting the run early.** Running it to a real CORE-prove length (2 h per-prove watchdog, `MEM=90G` systemd scope) shows the host RSS grows **monotonically**, ~1.5–2 GB/min, with no plateau:

```
19:48  13.7 GB  (CORE GKR entry)
20:06  44.6 GB
20:14  56.9 GB
20:23  70.6 GB
20:27  84.6 GB  → killed at the 90 GB scope cap, still climbing, CORE GKR not finished
```

Peak scales with wall-time (37 GB@14 min, 44 GB@27 min, 84.6 GB@41 min), so the "~37–44 GB" numbers were just early cutoffs. So **both paths balloon**:
- **marked** (`jit=True`): fuses the whole body into one program → all pyramid layers + round state live at once → ~104 GB fast → OOM.
- **markless** (`jit=False`): eager per-round host dispatch → per-round/per-layer state is **not released** → monotonic climb → would OOM too, just slower (and CORE GKR is also too slow to finish — >40 min unfinished).

Net: markless is not a workaround — it's slow **and** leaks. This reinforces the root cause: the GKR body's per-layer memory is never released (the fused `scan_build` holds all layers at once; the eager path accumulates them round-by-round). The fix still has to bound peak to `O(floor)` — release each layer as it's proved — not just change the dispatch mode.

**Comment by chokobole** [comment]
## The pyramid builder is not the dominant OOM source — the cause is `jit=True` tracing the whole LogUp-GKR body into one XLA program (→ #327)

Investigated on rsp `shard17` (`verify_prove_shard --max_stage=2`, selfbuilt jax + local xla). **Three independent, byte-correct changes do not move the OOM:**

1. **Builder — flat-buffer variant.** Replace the fixed-width `lax.scan` `ys` stack with a carried flat buffer (each transition writes its full-width plane at the layer offset; the next, smaller layer overwrites the dead tail). `jagged_circuit_test` 24/24. Still OOMs ~33 GB.
2. **Builder — shape-polymorphic `jax.export` host-loop** (the #327 pattern applied to the *build*). One symbolic-height transition binary, dispatched per layer at its true halving width; `scan_build_jagged_pyramid` becomes a host loop with no floor-width padding. `jagged_circuit_test` 24/24 + a new symbolic-export test (one binary, two heights, byte-identical). Still OOMs ~30–46 GB.
3. **Consumer.** `prove_logup_gkr_body` does `proved = [layers.pop() for _ in range(len(layers))]` then `ProveChain(JaggedGkrLayerRound(l) for l in proved)`. `ProveChain` is *designed* to release each proved layer before building the next ("at most one big-witness layer stays live"), but the materialized `proved` list pins the whole pyramid and defeats it. Replacing it with a destructive `layers.pop()`-per-yield generator restores the invariant (byte-identical, same floor-out order; sp1-zorch logup tests 40/40). Still OOMs.

### Root cause (A/B — only `prove_shard_chain(jit=...)` flipped)

- **`jit=True`** (production; the whole grind-free body traced into one XLA program): **OOM ≥46 GB.** XLA co-residents the widest layer's full-width sumcheck **and** every round's buffers in the single program.
- **`jit=False`** (eager → the sumcheck routes through the existing per-round `jax.export` host-loop, releasing each round): **completes**, peak host RSS **24.4 GB**, `zeta vs gpu_z_row` byte-match **PASS**, `chain byte-match ALL OK`. Slow (`[stage LogupGkrRound] 1050 s`, host-dispatch-bound) — so `jit=False` is the diagnosis + a full-prove correctness proof, not a production fix.

**Conclusion:** the OOM lives in *how the sumcheck is compiled* (one fused `@jit` co-residency), not in `scan_build_jagged_pyramid`. The builder's `O(N × plane_width)` stack is real and the shape-polymorphic rewrite above is a worthwhile standalone perf/memory improvement, but it is **not** the cure for this OOM. The fix belongs in **#327** (run the sumcheck as a host loop over cached per-round export binaries, decoupled from the outer `@jit`).

Suggest keeping this issue for the builder improvement and tracking the actual OOM fix on #327 (details there).
